File: compiler/extensions/java/runtime/src/zserio/runtime/io/BitStreamReader.java
package zserio.runtime.io;

import java.io.Closeable;
import java.io.IOException;
import java.math.BigInteger;

/**
 * An interface for a bit stream reader implementation.
 */
public interface BitStreamReader extends Closeable
{
    /**
     * Reads the next numBits as signed bit value.
     *
     * @param numBits Number of bits to read.
     *
     * @return Read signed bit value.
     *
     * @throws IOException If reading failed.
     */
    long readSignedBits(final int numBits) throws IOException;

    /**
     * Reads the next numBits of the bit stream as long value.
     *
     * @param numBits Number of bits to read.
     *
     * @return Read numBits value.
     *
     * @throws IOException If reading failed.
     */
    long readBits(final int numBits) throws IOException;

    /**
     * Reads the next signed byte (8 bits).
     *
     * @return Read signed byte value.
     *
     * @throws IOException If reading failed.
     */
    byte readByte() throws IOException;

    /**
     * Reads the next unsigned byte (8 bits) as short value.
     *
     * @return Read unsigned byte value.
     *
     * @throws IOException If reading failed.
     */
    short readUnsignedByte() throws IOException;

    /**
     * Reads the next signed short (16 bits).
     *
     * @return Read signed short value.
     *
     * @throws IOException If reading failed.
     */
    short readShort() throws IOException;

    /**
     * Read the next unsigned short (16 bits) as int value.
     *
     * @return Read unsigned short value.
     *
     * @throws IOException If reading failed.
     */
    int readUnsignedShort() throws IOException;

    /**
     * Reads the next signed integer (32 bits).
     *
     * @return Read signed integer value.
     *
     * @throws IOException If reading failed.
     */
    int readInt() throws IOException;

    /**
     * Reads the next unsigned integer (32 bits) as long value.
     *
     * @return Read unsigned integer value.
     *
     * @throws IOException If reading failed.
     */
    long readUnsignedInt() throws IOException;

    /**
     * Reads the next signed long (64 bits).
     *
     * @return Read signed long value.
     *
     * @throws IOException If reading failed.
     */
    long readLong() throws IOException;

    /**
     * Reads the next unsigned Big Integer value with containing numBits bits.
     *
     * @param numBits Number of bits to read.
     *
     * @return Read big integer with the given bits.
     *
     * @throws IOException If reading failed.
     */
    BigInteger readBigInteger(final int numBits) throws IOException;

    /**
     * Reads the next signed big Integer value with containing numBits bits.
     *
     * @param numBits Number of bits to read.
     *
     * @return Read big integer with the given bits.
     *
     * @throws IOException If reading failed.
     */
    BigInteger readSignedBigInteger(final int numBits) throws IOException;

    /**
     * Reads the a float16 value from the bit stream.
     *
     * @return Read half precision float value.
     *
     * @throws IOException If reading failed.
     */
    float readFloat16() throws IOException;

    /**
     * Reads the a float32 value from the bit stream.
     *
     * @return Read single precision float value.
     *
     * @throws IOException If reading failed.
     */
    float readFloat32() throws IOException;

    /**
     * Reads the a float64 value from the bit stream.
     *
     * @return Read double precision float value.
     *
     * @throws IOException If reading failed.
     */
    double readFloat64() throws IOException;

    /**
     * Reads a Zserio bytes from the bit stream.
     *
     * @return Read Zserio bytes as byte[].
     *
     * @throws IOException If reading failed.
     */
    byte[] readBytes() throws IOException;

    /**
     * Reads a Zserio string from the bit stream assuming it is encoded in UTF-8.
     *
     * @return Read Zserio string.
     *
     * @throws IOException If reading failed.
     */
    String readString() throws IOException;

    /**
     * Reads a Boolean value from the bit stream.
     *
     * @return Read boolean value.
     *
     * @throws IOException If reading failed.
     */
    boolean readBool() throws IOException;

    /**
     * Reads a variable 16 bit integer from the bit stream.
     *
     * @return Read VarInt16 value.
     *
     * @throws IOException If reading failed.
     */
    short readVarInt16() throws IOException;

    /**
     * Reads a variable 32 bit integer from the bit stream.
     *
     * @return Read VarInt32 value.
     *
     * @throws IOException If reading failed.
     */
    int readVarInt32() throws IOException;

    /**
     * Reads a variable 64 bit integer from the bit stream.
     *
     * @return Read VarInt64 value.
     *
     * @throws IOException If reading failed.
     */
    long readVarInt64() throws IOException;

    /**
     * Reads an unsigned variable 16 bit integer from the bit stream.
     *
     * @return Read unsigned VarUInt16 value.
     *
     * @throws IOException If reading failed.
     */
    short readVarUInt16() throws IOException;

    /**
     * Reads an unsigned variable 32 bit integer from the bit stream.
     *
     * @return Read unsigned VarUInt32 value.
     *
     * @throws IOException If reading failed.
     */
    int readVarUInt32() throws IOException;

    /**
     * Reads an unsigned variable 64 bit integer from the bit stream.
     *
     * @return Read unsigned VarUInt64 value.
     *
     * @throws IOException If reading failed.
     */
    long readVarUInt64() throws IOException;

    /**
     * Reads a signed variable integer from the bit stream.
     *
     * The integer takes up to 9 bytes to cover range &lt;-2^63, 2^63-1&gt;.
     *
     * @return Read signed VarInt value.
     *
     * @throws IOException If reading failed.
     */
    long readVarInt() throws IOException;

    /**
     * Reads an unsigned variable integer from the bit stream.
     *
     * The integer takes up to 9 bytes to cover range &lt;0, 2^64-1&gt;.
     *
     * @return Read unsigned VarUInt value.
     *
     * @throws IOException If reading failed.
     */
    BigInteger readVarUInt() throws IOException;

    /**
     * Reads a variable size integer from the bit stream.
     *
     * @return Read VarSize value.
     *
     * @throws IOException If reading failed.
     */
    int readVarSize() throws IOException;

    /**
     * Reads a bit buffer from the bit stream.
     *
     * @return Read bit buffer.
     *
     * @throws IOException If reading failed.
     */
    BitBuffer readBitBuffer() throws IOException;

    /**
     * Gets the current bit position.
     *
     * @return Current bit position counted from zero.
     */
    long getBitPosition();

    /**
     * Gets the current byte position.
     *
     * @return Current byte position counted from zero.
     */
    int getBytePosition();

    /**
     * Sets the bit position to the given value.
     *
     * @param position Bit position counted from zero to set.
     *
     * @throws IOException If the position cannot be set.
     */
    void setBitPosition(long position) throws IOException;

    /**
     * Aligns the bit position according to the aligning value.
     *
     * @param alignVal An aligning value to use.
     *
     * @throws IOException If the alignment failed.
     */
    void alignTo(final int alignVal) throws IOException;

    /**
     * Gets size of the underlying buffer in bits.
     *
     * @return Buffer bit size.
     */
    long getBufferBitSize();
}

File: compiler/extensions/python/runtime/src/zserio_cpp/BitStreamReader.cpp
#include "BitStreamReader.h"

namespace
{

std::unique_ptr<zserio::BitStreamReader> bitStreamReaderInit(const py::buffer_info& bufferInfo,
            py::object optionalBitSize)
{
    const uint8_t* buffer = static_cast<const uint8_t*>(bufferInfo.ptr);
    if (!optionalBitSize.is(py::none()))
    {
        const size_t bitSize = optionalBitSize.cast<size_t>();
        if ((bitSize + 7) / 8 > static_cast<size_t>(bufferInfo.size))
            throw zserio::CppRuntimeException("BitStreamReader: Bit size out of range for given buffer!");

        return std::make_unique<zserio::BitStreamReader>(buffer, bitSize, zserio::BitsTag());
    }
    else
    {
        return std::make_unique<zserio::BitStreamReader>(buffer, bufferInfo.size);
    }
}

} // namespace

namespace zserio_cpp
{

void pybindBitStreamReader(py::module_ module)
{
    py::class_<zserio::BitStreamReader>(module, "BitStreamReader")
            .def(py::init([](const py::object& buffer, py::object bitsize) {
                return bitStreamReaderInit(py::buffer(buffer).request(), bitsize);
            }), py::arg("buffer"), py::arg("bitsize") = py::none(), py::keep_alive<1, 2>())
            .def_static("from_bitbuffer", [](const zserio::BitBuffer& bitBuffer){
                return std::make_unique<zserio::BitStreamReader>(bitBuffer);
            }, py::keep_alive<0, 1>())
            .def("read_bits", [](zserio::BitStreamReader& self, py::int_ numbits){
                return self.readBits64(numbits.cast<uint8_t>());
            })
            .def("read_signed_bits", [](zserio::BitStreamReader& self, py::int_ numbits){
                return self.readSignedBits64(numbits.cast<uint8_t>());
            })
            .def("read_bits_unchecked", [](zserio::BitStreamReader& self, py::int_ numbits){
                return self.readBits64(numbits.cast<uint8_t>());
            })
            .def("read_signed_bits_unchecked", [](zserio::BitStreamReader& self, py::int_ numbits){
                return self.readSignedBits64(numbits.cast<uint8_t>());
            })
            .def("read_varint16", &zserio::BitStreamReader::readVarInt16)
            .def("read_varint32", &zserio::BitStreamReader::readVarInt32)
            .def("read_varint64", &zserio::BitStreamReader::readVarInt64)
            .def("read_varint", &zserio::BitStreamReader::readVarInt)
            .def("read_varuint16", &zserio::BitStreamReader::readVarUInt16)
            .def("read_varuint32", &zserio::BitStreamReader::readVarUInt32)
            .def("read_varuint64", &zserio::BitStreamReader::readVarUInt64)
            .def("read_varuint", &zserio::BitStreamReader::readVarUInt)
            .def("read_varsize", &zserio::BitStreamReader::readVarSize)
            .def("read_float16", &zserio::BitStreamReader::readFloat16)
            .def("read_float32", &zserio::BitStreamReader::readFloat32)
            .def("read_float64", &zserio::BitStreamReader::readFloat64)
            .def("read_bytes", [](zserio::BitStreamReader& self){
                auto bytes = self.readBytes();
                return py::bytearray(reinterpret_cast<char*>(bytes.data()), bytes.size());
            })
            .def("read_string", [](zserio::BitStreamReader& self){
                return self.readString();
            })
            .def("read_bool", &zserio::BitStreamReader::readBool)
            .def("read_bitbuffer", [](zserio::BitStreamReader& self) {
                return self.readBitBuffer();
            })
            .def("alignto", [](zserio::BitStreamReader& self, py::int_ alignment) {
                self.alignTo(alignment.cast<size_t>());
            })
            .def_property("bitposition",
                    &zserio::BitStreamReader::getBitPosition,
                    [](zserio::BitStreamReader& self, py::int_ bitposition) {
                        self.setBitPosition(bitposition.cast<zserio::BitStreamReader::BitPosType>());
                    })
            .def_property_readonly("buffer_bitsize",
                    &zserio::BitStreamReader::getBufferBitSize)
            ;
}

} // namespace zserio_cpp
